### PR TITLE
Use react-router; client- and server side rendering

### PR DESCRIPTION
**Pull request by okendoken**
Use react-router instead of react-routing library.

As per #345

**Update.** Here is a live version with react-router https://limitless-wave-36322.herokuapp.com/

---

This change is [Reviewable](https://reviewable.io/reviews/kriasoft/react-starter-kit/422)

### CODE REVIEW

**Comment by goldensunliu** [comment]
@okendoken There is a flash of HTML without CSS on initial render. I was able to reproduce locally

**Comment by okendoken** [comment]
@goldensunliu agreed, I noticed it too. Looking into it

**Comment by borovik96** [comment]
The same problem when using react-routing

**Comment by okendoken** [comment]
@goldensunliu @borovik96, updated, please check

**Comment by goldensunliu** [comment]
@okendoken  just curious what was the fix since you have squashed the commits

File: src/server.js [hunk 1]
```diff
@@ -34,23 +36,25 @@ server.use('/api/content', require('./api/content').default);
...
-      insertCss: styles => css.push(styles._getCss()),
-      onSetTitle: value => data.title = value,
-      onSetMeta: (key, value) => data[key] = value,
-      onPageNotFound: () => statusCode = 404,
-    };
-
-    await Router.dispatch({ path: req.path, query: req.query, context }, (state, component) => {
-      data.body = ReactDOM.renderToString(component);
+    match({ routes, location: req.url }, (error, redirectLocation, renderProps) => {
+      let statusCode = 200;
```
Comment (goldensunliu) [inline src/server.js]: does Redirect work?

Comment (goldensunliu) [inline src/server.js]: @okendoken I added redirect and error handling logic here: https://github.com/goldensunliu/react-starter-kit/commit/d6885b9a1d470e551229a03852f2bf051ed9fdcc just tested.

**Comment by okendoken** [comment]
@goldensunliu thanks for your comments
I tried to merge your commit and it seems that there are errors you need to fix:

```D:\projects\react-starter-kit\src\server.js
  44:13  error  `redirectPath` is never modified, use `const` instead  prefer-const
  46:15  error  Missing semicolon                                      semi
```
Could you please run `npm run test`, check for errors, fix them and make another commit (preferably squashing your commits into a single one)? It would be also nice if you force update your fork with last changes from mine - we have different commits history. Otherwise I will have to cherry-pick.

Thanks!

**Comment by goldensunliu** [comment]
@okendoken I will get on it

**Comment by goldensunliu** [comment]
@okendoken all set https://github.com/flatlogic/react-starter-kit/pull/2

**Comment by jeroenransijn** [comment]
Excited to see this! Thanks for the work everyone :)

**Comment by jeroenransijn** [comment]
I am trying out this fork/branch. I am trying to hook it up to redux, since that's what I am using. The issue that comes up is when I try to connect my component `export default connect(state => state)(withStyles(Navigation, s));`. The server just keeps loading, it get stuck on `ReactDOM.renderToString`. It's really difficult to debug this since nothing is returning. Any help is appreciated.

**Comment by okendoken** [comment]
@jeroenransijn I think you may try to check `getContextComponent` function https://github.com/kriasoft/react-starter-kit/pull/422/files#diff-3ca5e313cdf02368ce63986a36207c94R20

Sometime I observed the same behavior when there was no page found during component lookup. We probably need to rethrow the error from the inside of this function

**Comment by erichardson30** [comment]
I added the changes from https://github.com/kriasoft/react-starter-kit/compare/master...flatlogic:feature/react-router and am getting the following warning : Warning: React attempted to reuse markup in a container but the checksum was invalid. This generally means that you are using server rendering and the markup generated on the server was not what the client was expecting. React injected new markup to compensate which works but you have lost many of the benefits of server rendering. Instead, figure out why the markup being generated is different on the client or server:
 (client) <noscript data-reacti
 (server) <div data-reactid=".c

Anyone else seeing this or know why this is happening?

**Comment by okendoken** [comment]
@erichardson30 the fix was to use `match` function on the client side also. 

But where do you get it? I can't see it any issues. I have just redeployed herokuapp - https://limitless-wave-36322.herokuapp.com/. See nothing there...

**Comment by erichardson30** [comment]
@okendoken So when you run npm start, it tells you the server is running on localhost:5000 and it opens up the client on localhost:3000. The warning is being thrown when I access through localhost:3000. I went through https://github.com/kriasoft/react-starter-kit/pull/422/files#diff-3ca5e313cdf02368ce63986a36207c94R20 and copied all replaced all the files in my project with those.

**Comment by erichardson30** [comment]
@okendoken I ended up fixing the issue. In client.js I had to change 
`<Router {...renderProps} children={routes} history={Location} />`
to 
`<RouterContext {...renderProps} children={routes} history={Location} /> `
and instead of 
` import { match, Router } from 'react-router';`
I had to do
` import { match, RouterContext } from 'react-router';`

**Comment by erichardson30** [comment]
@okendoken where is the latest complete code with the react router so I can see what's wrong?

**Comment by okendoken** [comment]
@erichardson30 to get latest code with react-router changes you need to run 

    git clone https://github.com/flatlogic/react-starter-kit.git -b feature/react-router

this will create a new repo with `feacture/react-router branch`

Is there any way I can reproduce this issue locally?

**Comment by erichardson30** [comment]
@okendoken I was able to clone that branch, merge in my changes and everything seems to be good now. Thanks for all your help :)

**Comment by karlstanton** [comment]
+1 - would love to see this merged.

**Comment by okendoken** [comment]
synced with master

**Comment by erichardson30** [comment]
@okendoken Have you done anything using react-router where you implement using isomorphic rendering getting data from an external API?

**Comment by Bogdaan** [comment]
Great work, thanks!
If someone needs case authorization,
i merge this PR with some changes to my repo (https://github.com/Bogdaan/react-auth-kit/blob/master/src/server.js)

**Comment by Bogdaan** [comment]
@okendoken hi, you PR is incomplate
See https://github.com/flatlogic/react-starter-kit/blob/feature/react-router/tools/webpack.config.js
and client-side matching is incorrect

**Comment by okendoken** [comment]
@Bogdaan what's wrong with it?

**Comment by Bogdaan** [comment]
@okendoken for example see react-routing conditions in webpack.config.js (find by text match in ide).
In client.js at (https://github.com/flatlogic/react-starter-kit/blob/feature/react-router/src/client.js#L62) 
Why are you use `window.location` and `match` on client-side?

**Comment by okendoken** [comment]
@Bogdaan please see discussion above. In order to allow asynchronous routing on the client side you need to use `match` as well https://github.com/reactjs/react-router/issues/2714.

As of react-routing mentions, I'll remove them and update the PR, thanks for noticing.

**Comment by okendoken** [comment]
react-routing mentions removed. Commits rebased

**Comment by okendoken** [comment]
@koistya what's your feedback on this PR? Are you going to merge it? Please provide a resolution, as it's a bit annoying to keep the PR in sync merging changes from master. Thanks

**Comment by koistya** [comment]
@okendoken thanks for the PR! I'm currently working on the authentication feature, which should be done within next couple of days. Maybe I can take it over and do rebase and some modifications myself until it's ready for `master` branch?

**Comment by Bogdaan** [comment]
@okendoken looks like async routes  works fine without `render()` ( see https://github.com/reactjs/react-router/pull/2883 )

**Comment by okendoken** [comment]
@koistya thanks for your answer! Your timeline looks fine - I'll keep the fork in sync then. I was afraid this PR went unnoticed at all ;)

Just in case, I added you as a collaborator to our repo.

**Comment by FakhruddinAbdi** [comment]
I can't load and use react-bootstrap module using feature/react-router, while it's working fine with fresh install of RSK master branch

**Comment by stupidisum** [comment]
How is the progress?

**Comment by okendoken** [comment]
@stupidisum the code https://github.com/flatlogic/react-starter-kit/tree/feature/react-router was completely ready more than two month ago and we kept it in sync with react-starter-kit master branch for the whole month and asked project creator @koistya to provide a resolution on this PR. However, it seems that they are not really interested in introducing react-router to this repo and we won't be able to spend more resources on keeping this ignored PR in sync.

Feel free to use our repo and, if you like, keep it in sync with react-starter-kit.

Thanks!

Refs #160

**Comment by learntoswim** [comment]
@okendoken That's shame!

**Comment by stupidisum** [comment]
I'm struggling in doing this on my own these days.
This pr is OK for now.
But once authentication is added. Boom! Server-render problem again.
And today, I found [this](https://github.com/choonkending/react-webpack-node/issues/107) and they've been much more further than react-starter-kit.

**Comment by dhyegofernando** [comment]
I'm really waiting for this merge :+1:

**Comment by goldensunliu** [comment]
@dhyegofernando just use the fork for now. I don't think it will be merged into this project

**Comment by dhyegofernando** [comment]
@goldensunliu that's what I'm gonna do but it could be really useful if created in a separated branch just like `feature/redux`

**Comment by gardner** [comment]
It would be great to have the option to use react-router. The react-routing library doesn't meet our needs.

**Comment by okendoken** [comment]
A small update to this pr:

We forked RSK and created [react-dashboard](https://github.com/flatlogic/react-dashboard). It has react-router & bootstrap integrated. We plan to continuously maintain it and keep it in sync with RSK.

**Comment by ulani** [comment]
@deslee thank you very much for this PR! Unfortunately, we have close it due to inactivity. Feel free to re-open it or [join](https://discord.com/invite/2nKEnKq) our [Discord](https://discord.com/invite/2nKEnKq) channel for [discussion](https://github.com/kriasoft/react-starter-kit/discussions/1950).

NOTE: The `main` branch has been updated with React Starter Kit v2, using JAM-style architecture.

**Pull request closed without merging**